### ISSUE TITLE
Lot of files in tmp dir CFG_TMPDIR

### ISSUE THREAD

**Issue by dblommesteijn** [EUDAT-B2SHARE/b2share#741]
I've been uploading a massive amount of files into b2share, now the server we have is running out of space due to temp files. These are stored in `.virtualenvs/b2sahre/var/tmp` via config setting `CFG_TMPDIR`. This seems like a bad place to store tmp files, why not use `/tmp` as a default? And why are these temp files created in the first place?

**Comment by nharraud** [comment]
I don't know the original reason, but as I said in #740, Invenio has been designed so that by default everything is in the virtual environment you setup. One can change this behavior by overriding config variables.
This is not a bug as it works with a big enough disk, and even /tmp is limited in space. It is more a documentation issue.
After doing a grep of `CFG_TMPDIR` on Invenio I can see that it is used by many different modules, so I can't answer the "why" part. It is quite common to need temporary files.

As bibsched is quite slow to schedule bibupload tasks, the disk will fill with non processes submissions if you upload too fast. This should be fixed in the next Invenio release as bibupload will be removed.

**Comment by dblommesteijn** [comment]
@nharraud  "This is not a bug as it works with a big enough disk..."
I don't agree it's a bug, It wrote like 450Gb to the tmp space, causing the diskspace to run out and than caused a system crash. 

In what world is it normal to just keep on writing to tmp space?

**Comment by nharraud** [comment]
Well. The system crash is surely due to a conflict between the swapping configuration of the machine and the default configuration of B2Share. If you don't have a separate partition for `/tmp`, you will have the same effect by using it instead of `.virtualenvs/b2sahre/var/tmp`. Thus this is a configuration issue.

We can change the default if it makes more sense. I don't have any opinion about this yet.

If you ask any database to write data, it will do it as long as the disk is not full and then refuse/complain and/or crash. Filling a disk is not a bug.

**Comment by nharraud** [comment]
Also it is never possible to be sure that you will not fill the disk by writing a file. Another process might be writing on the disk just after you checked the remaining free space.

**Comment by nharraud** [comment]
We might send email alerts when disk space reaches configured thresholds. This would be a nice feature.
We might also refuse uploads when too many of them are waiting and/or take more time to send responses so that correctly implemented clients will wait before sending the next upload.

**Comment by nharraud** [comment]
One question I should have asked first, is inveniogc running on your server? It is the process cleaning the tmp files.

**Comment by nharraud** [comment]
I asked Jiri and it confirms what I thought. There are temporary files created for each not-yet-processed upload. As the bibtasks are scheduled slowly, the disk my be filled with arriving uploads. One way to make things faster is to upload by batch.

**Comment by tiborsimko** [comment]
>>  It wrote like 450Gb to the tmp space
>
> is inveniogc running on your server?

This is of utmost importance, because without garbage collection running, the files in temporary folders would stay there forever.  See `inveniogc --help` or see online e.g. [HOWTO Run guide](http://demo.invenio-software.org/help/admin/howto-run) that largely applies also to v2.x regarding the temporary files.

**Comment by dblommesteijn** [comment]
But why are these tmp files so large? Aren't they just copied from `var/tmp/ss` to the `var/data/files`data folder?

PS. when I run the garbage collector bibtask errors, because there are db tables missing:

```log
2015-08-19 15:41:59 -->
Traceback (most recent call last):
  File "/home/b2admin/b2share/invenio/legacy/bibsched/bibtask.py", line 615, in task_init
    ret = _task_run(task_run_fnc)
  File "/home/b2admin/b2share/invenio/legacy/bibsched/bibtask.py", line 1191, in _task_run
    if callable(task_run_fnc) and task_run_fnc():
  File "/home/b2admin/b2share/invenio/legacy/websession/inveniogc.py", line 666, in task_run_core
    guest_user_garbage_collector()
  File "/home/b2admin/b2share/invenio/legacy/websession/inveniogc.py", line 400, in guest_user_garbage_collector
    " session_expiry < %s """, (timelimit,))
  File "/home/b2admin/b2share/invenio/legacy/dbquery_mysql.py", line 268, in run_sql
    rc = cur.execute(sql, param)
  File "/usr/lib64/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib64/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
ProgrammingError: (1146, "Table 'invenio.session' doesn't exist")
2015-08-19 15:41:59 --> Unexpected error occurred: (1146, "Table 'invenio.session' doesn't exist").
2015-08-19 15:41:59 --> Traceback is:
2015-08-19 15:41:59 -->   File "/home/b2admin/b2share/invenio/legacy/bibsched/bibtask.py", line 615, in task_init
2015-08-19 15:41:59 -->     ret = _task_run(task_run_fnc)
2015-08-19 15:41:59 -->   File "/home/b2admin/b2share/invenio/legacy/bibsched/bibtask.py", line 1191, in _task_run
2015-08-19 15:41:59 -->     if callable(task_run_fnc) and task_run_fnc():
2015-08-19 15:41:59 -->   File "/home/b2admin/b2share/invenio/legacy/websession/inveniogc.py", line 666, in task_run_core
2015-08-19 15:41:59 -->     guest_user_garbage_collector()
2015-08-19 15:41:59 -->   File "/home/b2admin/b2share/invenio/legacy/websession/inveniogc.py", line 400, in guest_user_garbage_collector
2015-08-19 15:41:59 -->     " session_expiry < %s """, (timelimit,))
2015-08-19 15:41:59 -->   File "/home/b2admin/b2share/invenio/legacy/dbquery_mysql.py", line 268, in run_sql
2015-08-19 15:41:59 -->     rc = cur.execute(sql, param)
2015-08-19 15:41:59 -->   File "/usr/lib64/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
2015-08-19 15:41:59 -->     self.errorhandler(self, exc, value)
2015-08-19 15:41:59 -->   File "/usr/lib64/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
2015-08-19 15:41:59 -->     raise errorclass, errorvalue
2015-08-19 15:41:59 --> Exiting.
```

**Comment by nharraud** [comment]
The error looks like this issue inveniosoftware/invenio#3205 which was fixed a long time ago in Invenio maint-2.0 branch and was merged only recently in B2Share (When I merge Invenio maintenance branch). Inveniogc is looking for the `session` table which has been moved from mysql to redis.
Upgrading to 1.6 should do the trick...once it is released. Or we can make an hotfix in one of the  maintenance branches. What version of B2Share are you using?

Regarding the size of the tmp files, I don't know, I can ask for more information when I have some time.

**Comment by tiborsimko** [comment]
The files are copied from the temporary place to the destination, but they are also left in that place, since in general anyone could upload a file from anywhere.    They are not removed after upload.

The garbage collector has several options, it attempts to clean sessions, which is not applicable in the case you are using Redis for sessions.  (I'll check the defaults in v2.0.)  Perhaps `inveniogc -p` is enough in your case, this cleans only temporary files.  Note though that it deletes only files older than N1 days and compresses files newer than N2 days; see  variables like `CFG_MAX_ATIME_RM_LOG` in `inveniogc.py`.  If you upload many files in a short time, manual cleaning may be wanted.

**Comment by tiborsimko** [comment]
> I'll check the defaults in v2.0.

@nharraud has been quicker :smile: The `inveniogc` default argument problem was indeed fixed in Invenio v2.0.5 that was [released](https://github.com/inveniosoftware/invenio/releases/tag/v2.0.5) on July 17th.  (Separating B2SHARE repository overlay from Invenio core will help in upgrading.)

**Comment by dblommesteijn** [comment]
I've enabled the garbage collector and the `tmp/` folder keeps on filling with files. So it's not working, any ideas?

**Comment by tiborsimko** [comment]
As I mentioned above, note that `inveniogc` deletes only files older than N1 days and compresses files newer than N2 days.  If you are populating them faster, you may need (1) either to reduce those configuration days in the script, or (2) you can delete the temporary files manually after they have been successfully uploaded.  (E.g. when the bibsched queue is fully empty and there are no more incoming tasks.)  

(I assume the temporary space problems  are created by files looking like `var/tmp/bibdocfile__QYmzX.gif`, i.e. files coming from document upload.  If not, please send `ls -l` of the files that consume the most space in that place...)

**Comment by dblommesteijn** [comment]
@tiborsimko they look like the file types you've just described. So basically there is no automatic solution... 
Can we disable the temp space directory, and copy directly to the data directory?

**Comment by tiborsimko** [comment]
> 200 uploads of 2Gb

If the files are uploaded from the same local disk to the same local disk, passing by the temporary space, this means that the box might roughly ~3 x 200 x 2 GB = ~1.2 TB of disk space available, roughly speaking.

> So basically there is no automatic solution. 

One technique would be to use the `--post-process` option of `bibupload` that accepts a Python script (=bibtasklet) that can do anything, e.g. remove file from the temporary place.  Invenio comes with `bst_send_email()` that sends email after a successful upload.

Another technique is to use `inveniogc` but reduce those parameters I was talking about (e.g. `CFG_MAX_ATIME_RM_BIBDOC` to 1).  This could be used if the files are coming in sparse bunches. 

(If you have them all upfront, then it's better to either script uploading and cleaning, or use the post processing option to bibupload mentioned above.)

> Can we disable the temp space directory, and copy directly to the data directory?

It would be possible to upload small "placeholder" files using the same names, and then copy over the "real files" to their destination places manually, followed by launching ``bibdocfile --fix-all`` job in order to recalculate MD5 checksums etc.

**Comment by emanueldima** [comment]
Should be fixed even in B2SHARE 1.6, but we'll check specifically in v2.

**Comment by emanueldima** [comment]
not valid in v2
